pkg/metricsadapter/controller.go

```diff
@@ -0,0 +1,166 @@
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "metrics-adapter"),
+	}
+	controller.addEventHandler()
+
+	return controller
+}
+
+// addEventHandler adds event handler for cluster
+func (m *MetricsController) addEventHandler() {
+	clusterInformer := m.InformerFactory.Cluster().V1alpha1().Clusters().Informer()
+	_, err := clusterInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    m.addCluster,
+		UpdateFunc: m.updateCluster,
+		DeleteFunc: m.deleteCluster,
+	})
+	if err != nil {
+		klog.Errorf("Failed to add cluster event handler for cluster: %v", err)
+	}
+}
+
+// addCluster adds cluster to queue
+func (m *MetricsController) addCluster(obj interface{}) {
+	cluster := obj.(*clusterV1alpha1.Cluster)
+	m.queue.Add(cluster.GetName())
+}
+
+// updateCluster updates cluster in queue
+func (m *MetricsController) updateCluster(oldObj, curObj interface{}) {
+	curCluster := curObj.(*clusterV1alpha1.Cluster)
+	oldCluster := oldObj.(*clusterV1alpha1.Cluster)
+	if curCluster.ResourceVersion == oldCluster.ResourceVersion {
+		// no change, do nothing.
+		return
+	}
+
+	if curCluster.DeletionTimestamp != nil {
+		// cluster is being deleted.
+		m.queue.Add(curCluster.GetName())
+	}
+
+	if !reflect.DeepEqual(curCluster.Spec, oldCluster.Spec) {
+		// Cluster.Spec is changed, rebuild informer.
+		m.InformerManager.Stop(curCluster.GetName())
+		m.queue.Add(curCluster.GetName())
+	}
+}
+
+// deleteCluster deletes cluster in queue
+func (m *MetricsController) deleteCluster(obj interface{}) {
+	cluster := obj.(*clusterV1alpha1.Cluster)
+	m.queue.Add(cluster.GetName())
+}
+
+// startController starts controller
+func (m *MetricsController) startController(stopCh <-chan struct{}) {
+	m.InformerFactory.WaitForCacheSync(stopCh)
+
+	go wait.Until(m.worker, time.Second, stopCh)
+
+	go func() {
+		<-stopCh
+		genericmanager.StopInstance()
+		klog.Infof("Shutting down karmada-metrics-adapter")
+	}()
+}
+
+// worker is a worker for handle the data in queue
+func (m *MetricsController) worker() {
+	for m.handleClusters() {
+	}
+}
+
+// handleClusters handles clusters changes
+func (m *MetricsController) handleClusters() bool {
+	key, shutdown := m.queue.Get()
+	if shutdown {
+		klog.Errorf("Fail to pop item from queue")
+		return false
+	}
+	defer m.queue.Done(key)
+
+	clusterName := key.(string)
+	cls, err := m.ClusterLister.Get(clusterName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("try to stop cluster informer %s", clusterName)
+			m.InformerManager.Stop(clusterName)
+			return true
+		}
+		return false
+	}
+
+	if !cls.DeletionTimestamp.IsZero() {
+		klog.Infof("try to stop cluster informer %s", clusterName)
+		m.InformerManager.Stop(clusterName)
+		return true
+	}
+
+	if !util.IsClusterReady(&cls.Status) {
+		klog.Warningf("cluster %s is notReady try to stop this cluster informer", clusterName)
+		m.InformerManager.Stop(clusterName)
+		return false
+	}
```

Comment:
I add this check `util.IsClusterReady(&curCluster.Status) != util.IsClusterReady(&oldCluster.Status)`, if the ready status changed, it will trigger the restored.